rpc-broker/src/policy.c

```diff
@@ -0,0 +1,290 @@
+            DBUS_BROKER_WARNING("Unrecognized Rule-Token: %s", token);
+            free_rule(*current);
+            current = NULL;
+            return -1;
+        }
+
+        token = strtok_r(NULL, delimiter, &ruleptr);
+    }
+
+    return 0;
+}
+
+static inline void get_rules(DBusConnection *conn, struct domain_policy *dom)
+{
+    int rule_idx;
+
+    for (rule_idx=0; rule_idx < MAX_RULES; rule_idx++) {
+        char *arg;
+        DBUS_REQ_ARG(arg, "/vm/%s/rpc-firewall-rules/%d",
+                     dom->uuid, rule_idx);
+
+        char *rulestring = db_query(conn, arg);
+
+        free(arg);
+
+        if (!rulestring)
+            break;
+
+        struct rule *policy_rule = &(dom->rules[rule_idx]);
+
+        if (create_rule(policy_rule, rulestring) < 0)
+            free_rule(*policy_rule);
+        else
+            dom->count++;
+
+        free(rulestring);
+    }
+
+}
+
+static void build_etc_policy(struct etc_policy *etc, const char *rule_filepath)
+{
+    struct stat policy_stat;
+    etc->count = 0;
+
+    if (stat(rule_filepath, &policy_stat) < 0) {
+        DBUS_BROKER_WARNING("/etc policy stat of file <%s> failed %s",
+                             rule_filepath, strerror(errno));
+        return;
+    }
+
+    char *filename = basename(rule_filepath);
+
+    if (filename[0] == '\0') {
+        DBUS_BROKER_WARNING("/etc policy invalid file <%s>", rule_filepath);
+        return;
+    }
+
+    etc->filename = filename;
+
+    size_t policy_size = policy_stat.st_size;
+
+    // There is no officially defined limit on policy file size.
+    // Rpc-broker defines a buffer size and for now limit based on that size.
+    if (policy_size > ETC_MAX_FILE) {
+        DBUS_BROKER_WARNING("/etc policy file %s exceeds buffer size <%zu>",
+                             rule_filepath, policy_size);
+        return;
+    }
+
+    int policy_fd = open(rule_filepath, O_RDONLY);
+
+    if (policy_fd < 0) {
+        DBUS_BROKER_WARNING("/etc policy file %s failed to open %s",
+                              rule_filepath, strerror(errno));
+        return;
+    }
+
+    int rbytes = read(policy_fd, etc->etc_file, policy_size);
+
+    if (rbytes < 0) {
+        close(policy_fd);
+        DBUS_BROKER_WARNING("/etc policy file %s invalid read %s",
+                              rule_filepath, strerror(errno));
+        return;
+    }
+
+    etc->etc_file[rbytes] = '\0';
+
+    const char *newline = "\n";
+    char *fileptr;
+    char *rule_token = strtok_r(etc->etc_file, newline, &fileptr);
+    int idx = 0;
+
+    while (rule_token &&  idx < MAX_RULES) {
+        if (isalpha(rule_token[0])) {
+            char *line = strdup(rule_token);
+            struct rule *current = &(etc->rules[idx]);
+            create_rule(current, line) < 0 ? free_rule(*current) : idx++;
+            free(line);
+        }
+
+        rule_token = strtok_r(NULL, newline, &fileptr);
+    }
+
+    etc->count = idx;
+    close(policy_fd);
+}
+
+DBusMessage *db_list(void)
```

Comment:
This function has nothing to do with policy and should be together with db_query in the same file.